org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTConverter.java

```diff
@@ -3631,14 +3631,33 @@ public TypeParameter convert(org.eclipse.jdt.internal.compiler.ast.TypeParameter
 	}
 
 	public Pattern convert(org.eclipse.jdt.internal.compiler.ast.TypePattern pattern) {
-		TypePattern typePattern = new TypePattern(this.ast);
+		TypePattern typePattern;
+		if (pattern.local == null) {
+			return createFakeNullPattern(pattern);
+		}
+		if(this.ast.apiLevel < AST.JLS22_INTERNAL) {
+			typePattern = new TypePattern(this.ast);
+		} else {
+			if(pattern.local != null && !(String.valueOf(pattern.local.name[0]) == "_") && pattern.local.type != null) { //$NON-NLS-1$
```

Comment:
`==` is most likely wrong here

Reply:
Done.

Reply:
Did you see my suggestion regarding comparing char instead of String?

Creating a 1 char string just for the purpose of this comparison is wasteful.

Reply:
Okay @stephan-herrmann, I was interpreted you suggestion in a different way and sorry for the confusion. Anyway, I have modified the code.

Reply:
@subyssurendran666 you correctly reacted to the bug I mentioned. Using `==` for strings is (generally) wrong, and `equals()` is better.

But improving code quality may involve more than one step :) -- so after I wrote my initial observation I looked deeper and saw that using strings wasn't even necessary. So you could have taken a short cut: instead of replacing `==` with `equals()` you could have gone directly to the `char` comparison. With this fix, code is not only correct, but also more efficient.

Now, we are at the optional step of polishing code style: what is the benefit of introducing a new variable `char1`? Perhaps it helped you realize that `'_'` is indeed of type `char` (as opposed to `"_"`), but for future readers this will only be a (small) distraction :)

Reply:
Stephan, I have modified the code. I changed it to use character comparison, i.e., pattern.local.name[0] == '_'.  I hope this code is fine.

Reply:
@stephan-herrmann Could you please approve my PR if you are okay with the current code?

Reply:
As mentioned "I didn't plan to do a full review" so I just removed the review request to signal that I don't see problems any more (but I didn't look at the change in its entirety).